src/tir/transforms/storage_flatten.cc

```diff
@@ -50,6 +50,933 @@ using runtime::StorageRank;
+      }
+      remap.begins = std::move(mapped_begins);
+    }
+
+    ICHECK(target_info.remap == nullptr) << "Indirect remapping not handled";
+
+    for (size_t i = 0; i < remap.begins.size(); i++) {
+      remap.begins.Set(i, bound_analyzer_->Simplify(remap.begins[i]));
+      remap.extents.Set(i, bound_analyzer_->Simplify(remap.extents[i]));
+    }
+
+    // Add a buffer remap entry
+    {
+      BufferEntry source_info;
+      source_info.buffer = source;
+      source_info.remap = std::make_unique<RemapInfo>(remap);
+
+      buf_map_[source.get()] = std::move(source_info);
+    }
+
+    // Define remappings of any remaining Variables (e.g. Store/Load nodes).
+    ArgBinder binder(&var_remap_);
+
+    // Define a view that represents the source's view into the target
+    // buffer.  This Buffer object is only used to define the mapping
+    // to the target buffer, and never actually appears in the TIR
+    // graph.
+    Buffer view = remap.target.MakeSlice(remap.begins, remap.extents);
+    if (source->strides.size() == 0) {
+      ICHECK_EQ(view->strides.size(), 0U)
+          << "Cannot bind a compact buffer to a strided buffer" << view->strides;
+    } else {
+      // Add explicit strides to the view, in order to bind to source.strides[i].
+      view = view.MakeStrideView();
+    }
+    binder.BindBuffer(source, view, source->name, true);
+
+    // Apply the remaps
+    Stmt body = op->body;
+    body = MergeNest(binder.asserts(), body);
+    body = MergeNest(binder.init_nest(), body);
+    body = this->VisitStmt(body);
+    // remove the binds
+    for (const Var& v : binder.defs()) {
+      var_remap_.erase(v.get());
+    }
+    return body;
+  }
+
+  struct RemapInfo {
+    Buffer target;
+    Array<PrimExpr> begins;
+    Array<PrimExpr> extents;
+  };
+
+  // The buffer entry in the flatten map
+  struct BufferEntry {
+    // The storage buffer
+    Buffer buffer;
+    // the bounds of realization, can be null, means everything
+    Region bounds;
+    // Whether the buffer is external
+    bool external{false};
+    // Whether we are within the allocation scope of the buffer.
+    bool in_scope{true};
+
+    // The buffer to which the storage buffer should be remapped.
+    std::unique_ptr<RemapInfo> remap{nullptr};
+
+    PrimExpr ElemOffset() const {
+      ICHECK(remap);
+
+      Buffer copy = remap->target;
+      {
+        Array<PrimExpr> shape;
+        for (auto r : bounds) {
+          shape.push_back(r->extent);
+        }
+        copy.CopyOnWrite()->shape = std::move(shape);
+      }
+
+      Buffer target_slice = copy.MakeSlice(remap->begins, remap->extents);
+      if (buffer->strides.size() == 0) {
+        ICHECK_EQ(target_slice->strides.size(), 0U)
+            << "Trying to bind compact buffer to strided one strides=" << target_slice->strides;
+      } else {
+        target_slice = target_slice.MakeStrideView();
+      }
+
+      return copy->ElemOffset(remap->begins);
+    }
+  };
```

Comment:
Not seeing `struct BufferEntry::ElemOffset` used anywhere. Consider removing or refactoring to use the method?

Reply:
And removed.  It was part of an earlier (and broken) implementation.